test_decorator failure for Python 3.11.4

**Describe the bug**
test_decorator.py failed during collection for Python 3.11.4 at line 65.
```python
>> signature = signature.replace(parameters=new_parameters)  
```
```bash
test/test_decorator.py:86: in <module>
    ("--b=2", 4, partial(_fn_with_all_argument_types, 1, c=1)),
test/test_decorator.py:66: in partial
    signature = signature.replace(parameters=new_parameters)
conda/envs/py311test/lib/python3.11/inspect.py:3052: in replace
    return type(self)(parameters,
.conda/envs/py311test/lib/python3.11/inspect.py:3008: in __init__
    raise ValueError(msg)
E   ValueError: non-default argument follows default argument
---------------------------------------------------------------
```
The test case ```("--b=2", 4, partial(_fn_with_all_argument_types, 1, c=1)),``` has parameters ```[<Parameter "a: int = 1">, <Parameter "b: int">, <Parameter "c: int = 1">]```


**To Reproduce**
```
pytest test/test_decorator.py
```

**Desktop (please complete the following information):**
 - Version: master
 - Python version: 3.11.4

**Additional context**
Add any other context about the problem here.

--- a/test/test_decorator.py
+++ b/test/test_decorator.py
@@ -41,15 +41,7 @@ class ListWithPopDefault(collections.UserList):
             return default_value
 
 
-def delay_evaluation_wrapper(fn: Callable) -> Callable:
-    def delayed_call(*args, **kwargs):
-        return lambda: fn(*args, **kwargs)
-
-    return delayed_call
-
-
-@delay_evaluation_wrapper
-def partial(fn: Callable, *args, **kwargs) -> Callable:
+def change_defaults(fn: Callable, *args, **kwargs) -> Callable:
     """Partial via changing the signature defaults."""
 
     @functools.wraps(fn)
@@ -81,35 +73,43 @@ def _xfail_in_py311(*param):
         marks=pytest.mark.xfail(
             sys.version_info >= (3, 11),
             reason="TODO: test doesn't work in Python 3.11",
+            raises=ValueError,  # "non-default argument follows default argument"
             strict=True,
         ),
     )
 
 
 @pytest.mark.parametrize(
-    "args, expected, delay_wrapper",
+    "args, expected, fn",
     [
-        ("", 1, partial(_fn_with_positional_only, 1)),
-        ("2", 2, partial(_fn_with_positional_only, 1)),
-        ("2", 2, partial(_fn_with_positional_only)),
-        ("", 1, partial(_fn_with_keyword_only, x=1)),
-        ("--x=2", 2, partial(_fn_with_keyword_only, x=1)),
-        ("--x=2", 2, partial(_fn_with_keyword_only)),
-        ("", 3, partial(_fn_with_all_argument_types, 1, b=1, c=1)),
-        ("2", 4, partial(_fn_with_all_argument_types, b=1, c=1)),
-        ("2 --b=2", 5, partial(_fn_with_all_argument_types, c=1)),
-        ("2 --b=2 --c=2", 6, partial(_fn_with_all_argument_types)),
-        ("--c=2", 4, partial(_fn_with_all_argument_types, 1, b=1)),
-        _xfail_in_py311("--b=2", 4, partial(_fn_with_all_argument_types, 1, c=1)),
-        _xfail_in_py311("--b=2 --c=2", 5, partial(_fn_with_all_argument_types, 1)),
+        ("", 1, change_defaults(_fn_with_positional_only, 1)),
+        ("2", 2, change_defaults(_fn_with_positional_only, 1)),
+        ("2", 2, change_defaults(_fn_with_positional_only)),
+        ("", 1, change_defaults(_fn_with_keyword_only, x=1)),
+        ("--x=2", 2, change_defaults(_fn_with_keyword_only, x=1)),
+        ("--x=2", 2, change_defaults(_fn_with_keyword_only)),
+        ("", 3, change_defaults(_fn_with_all_argument_types, 1, b=1, c=1)),
+        ("2", 4, change_defaults(_fn_with_all_argument_types, b=1, c=1)),
+        ("2 --b=2", 5, change_defaults(_fn_with_all_argument_types, c=1)),
+        ("2 --b=2 --c=2", 6, change_defaults(_fn_with_all_argument_types)),
+        ("--c=2", 4, change_defaults(_fn_with_all_argument_types, 1, b=1)),
+        _xfail_in_py311(
+            "--b=2", 4, functools.partial(change_defaults, _fn_with_all_argument_types, 1, c=1)
+        ),
+        _xfail_in_py311(
+            "--b=2 --c=2", 5, functools.partial(change_defaults, _fn_with_all_argument_types, 1)
+        ),
     ],
 )
 def test_simple_arguments(
     args: str,
     expected: int,
-    delay_wrapper: Callable,
+    fn: Callable,
 ):
-    fn = delay_wrapper()
+    if isinstance(fn, functools.partial):
+        # In Python 3.11.4, the inspect module had a backward-incompatible change. We need to have
+        # this additional level of indirection.
+        fn = fn()
     decorated = sp.decorators.main(fn, args=args)
     assert decorated() == expected
 
@@ -119,20 +119,21 @@ def _fn_with_nested_dataclass(x: int, /, *, data: AddThreeNumbers) -> int:
 
 
 @pytest.mark.parametrize(
-    "args, expected, delay_wrapper",
+    ("args", "expected", "fn"),
     [
-        _xfail_in_py311("", 1, partial(_fn_with_nested_dataclass, 1, data=AddThreeNumbers())),
-        ("--a=1", 2, partial(_fn_with_nested_dataclass, 1)),
-        ("--a=1 --b=1", 3, partial(_fn_with_nested_dataclass, 1)),
-        ("--a=1 --b=1 --c=1", 4, partial(_fn_with_nested_dataclass, 1)),
-        ("2 --a=1 --b=1 --c=1", 5, partial(_fn_with_nested_dataclass)),
+        _xfail_in_py311(
+            "", 1, change_defaults(_fn_with_nested_dataclass, 1, data=AddThreeNumbers())
+        ),
+        ("--a=1", 2, change_defaults(_fn_with_nested_dataclass, 1)),
+        ("--a=1 --b=1", 3, change_defaults(_fn_with_nested_dataclass, 1)),
+        ("--a=1 --b=1 --c=1", 4, change_defaults(_fn_with_nested_dataclass, 1)),
+        ("2 --a=1 --b=1 --c=1", 5, change_defaults(_fn_with_nested_dataclass)),
     ],
 )
 def test_nested_dataclass(
     args: str,
     expected: int,
-    delay_wrapper: Callable,
+    fn: Callable,
 ):
-    fn = delay_wrapper()
     decorated = sp.decorators.main(fn, args=args)
     assert decorated() == expected

--- a/test/test_decorator.py
+++ b/test/test_decorator.py
@@ -44,27 +44,8 @@ class ListWithPopDefault(collections.UserList):
 
 
 def partial(fn: Callable, *args, **kwargs) -> Callable:
-    """Partial via changing the signature defaults."""
-
-    @functools.wraps(fn)
-    def _wrapper(*other_args, **other_kwargs):
-        return fn(*other_args, **other_kwargs)
-
-    signature = inspect.signature(fn)
-    parameters = signature.parameters
-
-    args = ListWithPopDefault(args)
-
-    new_parameters = []
-    for key, param in parameters.items():
-        param = typing.cast(inspect.Parameter, param)
-        default = args.pop(0, None) or kwargs.get(key, None)
-        if default is not None:
-            param = param.replace(default=default)
-        new_parameters.append(param)
-
-    signature = signature.replace(parameters=new_parameters)
-    _wrapper.__signature__ = signature
+    _wrapper = functools.partial(fn, *args, **kwargs)
+    _wrapper.__qualname__ = fn.__qualname__
 
     return _wrapper
 
@@ -73,7 +54,7 @@ def partial(fn: Callable, *args, **kwargs) -> Callable:
     "args, expected, fn",
     [
         ("", 1, partial(_fn_with_positional_only, 1)),
-        ("2", 2, partial(_fn_with_positional_only, 1)),
+        ("2", 2, partial(_fn_with_positional_only)),
         ("2", 2, _fn_with_positional_only),
         ("", 1, partial(_fn_with_keyword_only, x=1)),
         ("--x=2", 2, partial(_fn_with_keyword_only, x=1)),

--- a/test/test_decorator.py
+++ b/test/test_decorator.py
@@ -43,7 +43,8 @@ class ListWithPopDefault(collections.UserList):
 
 def delay_evaluation_wrapper(fn: Callable) -> Callable:
     def delayed_call(*args, **kwargs):
-        return lambda : fn(*args, **kwargs)
+        return lambda: fn(*args, **kwargs)
+
     return delayed_call
 
 
@@ -115,9 +116,6 @@ def test_simple_arguments(
 
 def _fn_with_nested_dataclass(x: int, /, *, data: AddThreeNumbers) -> int:
     return x + data()
-
-
-
 
 
 @pytest.mark.parametrize(

--- a/test/test_decorator.py
+++ b/test/test_decorator.py
@@ -32,14 +32,6 @@ def _fn_with_all_argument_types(a: int, /, b: int, *, c: int) -> int:
     return a + b + c
 
 
-class ListWithPopDefault(collections.UserList):
-    def pop(self, index, default_value=None):
-        try:
-            return super().pop(index)
-        except IndexError:
-            return default_value
-
-
 def partial(fn: Callable, *args, **kwargs) -> Callable:
     _wrapper = functools.partial(fn, *args, **kwargs)
     _wrapper.__qualname__ = fn.__qualname__

--- a/test/test_decorator.py
+++ b/test/test_decorator.py
@@ -4,7 +4,8 @@ import dataclasses
 import functools
 import sys
 from typing import Callable
-
+import typing
+import inspect
 import pytest
 
 import simple_parsing as sp
@@ -32,42 +33,45 @@ def _fn_with_all_argument_types(a: int, /, b: int, *, c: int) -> int:
     return a + b + c
 
 
+class ListWithPopDefault(collections.UserList):
+    def pop(self, index, default_value=None):
+        try:
+            return super().pop(index)
+        except IndexError:
+            return default_value
+
+
+def delay_evaluation_wrapper(fn: Callable) -> Callable:
+    def delayed_call(*args, **kwargs):
+        return lambda : fn(*args, **kwargs)
+    return delayed_call
+
+
+@delay_evaluation_wrapper
 def partial(fn: Callable, *args, **kwargs) -> Callable:
-    _wrapper = functools.partial(fn, *args, **kwargs)
-    _wrapper.__qualname__ = fn.__qualname__
+    """Partial via changing the signature defaults."""
+
+    @functools.wraps(fn)
+    def _wrapper(*other_args, **other_kwargs):
+        return fn(*other_args, **other_kwargs)
+
+    signature = inspect.signature(fn)
+    parameters = signature.parameters
+
+    args = ListWithPopDefault(args)
+
+    new_parameters = []
+    for key, param in parameters.items():
+        param = typing.cast(inspect.Parameter, param)
+        default = args.pop(0, None) or kwargs.get(key, None)
+        if default is not None:
+            param = param.replace(default=default)
+        new_parameters.append(param)
+
+    signature = signature.replace(parameters=new_parameters)
+    _wrapper.__signature__ = signature
 
     return _wrapper
-
-
-@pytest.mark.parametrize(
-    "args, expected, fn",
-    [
-        ("", 1, partial(_fn_with_positional_only, 1)),
-        ("2", 2, partial(_fn_with_positional_only)),
-        ("2", 2, _fn_with_positional_only),
-        ("", 1, partial(_fn_with_keyword_only, x=1)),
-        ("--x=2", 2, partial(_fn_with_keyword_only, x=1)),
-        ("--x=2", 2, _fn_with_keyword_only),
-        ("", 3, partial(_fn_with_all_argument_types, 1, b=1, c=1)),
-        ("2", 4, partial(_fn_with_all_argument_types, b=1, c=1)),
-        ("2 --b=2", 5, partial(_fn_with_all_argument_types, c=1)),
-        ("2 --b=2 --c=2", 6, _fn_with_all_argument_types),
-        ("--c=2", 4, partial(_fn_with_all_argument_types, 1, b=1)),
-        ("--b=2", 4, partial(_fn_with_all_argument_types, 1, c=1)),
-        ("--b=2 --c=2", 5, partial(_fn_with_all_argument_types, 1)),
-    ],
-)
-def test_simple_arguments(
-    args: str,
-    expected: int,
-    fn: Callable,
-):
-    decorated = sp.decorators.main(fn, args=args)
-    assert decorated() == expected
-
-
-def _fn_with_nested_dataclass(x: int, /, *, data: AddThreeNumbers) -> int:
-    return x + data()
 
 
 def _xfail_in_py311(*param):
@@ -82,7 +86,42 @@ def _xfail_in_py311(*param):
 
 
 @pytest.mark.parametrize(
-    "args, expected, fn",
+    "args, expected, delay_wrapper",
+    [
+        ("", 1, partial(_fn_with_positional_only, 1)),
+        ("2", 2, partial(_fn_with_positional_only, 1)),
+        ("2", 2, partial(_fn_with_positional_only)),
+        ("", 1, partial(_fn_with_keyword_only, x=1)),
+        ("--x=2", 2, partial(_fn_with_keyword_only, x=1)),
+        ("--x=2", 2, partial(_fn_with_keyword_only)),
+        ("", 3, partial(_fn_with_all_argument_types, 1, b=1, c=1)),
+        ("2", 4, partial(_fn_with_all_argument_types, b=1, c=1)),
+        ("2 --b=2", 5, partial(_fn_with_all_argument_types, c=1)),
+        ("2 --b=2 --c=2", 6, partial(_fn_with_all_argument_types)),
+        ("--c=2", 4, partial(_fn_with_all_argument_types, 1, b=1)),
+        _xfail_in_py311("--b=2", 4, partial(_fn_with_all_argument_types, 1, c=1)),
+        _xfail_in_py311("--b=2 --c=2", 5, partial(_fn_with_all_argument_types, 1)),
+    ],
+)
+def test_simple_arguments(
+    args: str,
+    expected: int,
+    delay_wrapper: Callable,
+):
+    fn = delay_wrapper()
+    decorated = sp.decorators.main(fn, args=args)
+    assert decorated() == expected
+
+
+def _fn_with_nested_dataclass(x: int, /, *, data: AddThreeNumbers) -> int:
+    return x + data()
+
+
+
+
+
+@pytest.mark.parametrize(
+    "args, expected, delay_wrapper",
     [
         _xfail_in_py311("", 1, partial(_fn_with_nested_dataclass, 1, data=AddThreeNumbers())),
         ("--a=1", 2, partial(_fn_with_nested_dataclass, 1)),
@@ -94,7 +133,8 @@ def _xfail_in_py311(*param):
 def test_nested_dataclass(
     args: str,
     expected: int,
-    fn: Callable,
+    delay_wrapper: Callable,
 ):
+    fn = delay_wrapper()
     decorated = sp.decorators.main(fn, args=args)
     assert decorated() == expected

--- a/test/test_decorator.py
+++ b/test/test_decorator.py
@@ -1,11 +1,8 @@
-"""Tests simple parsing decorators.
-"""
+"""Tests simple parsing decorators."""
 import collections
 import dataclasses
 import functools
-import inspect
 import sys
-import typing
 from typing import Callable
 
 import pytest

